# Guided Capstone Step 4 - Pre-processing and Training Data Development

### **The Data Science Method**  


1.   Problem Identification 

2.   Data Wrangling 
  
3.   Exploratory Data Analysis 
 
4.   **Pre-processing and Training Data Development**

5.  Modeling
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

In this step of the guided capstone, you'll revisit some of the steps you completed to solve questions in step 4. Revisiting these steps will help you get additional practice completing this kind of work.

**<font color='teal'> Start by loading the necessary packages, as you've done in the previous steps. Print out your current working directory to confirm that you are in the correct project directory. </font>**

In [0]:
#%reset
%reset_selective -f regex
import os
import pandas as pd
import datetime
import seaborn as sns
import matplotlib.pyplot as plt
import pprint
import numpy as np
import seaborn as sns
from scipy.stats import zscore
from sklearn import preprocessing
%matplotlib inline

**<font color='teal'>  Load the csv file you created in step 3. Remember, it should be saved inside your data subfolder. Print the first five rows.</font>**

In [456]:
df=pd.read_csv('step3b_output_withunnamed.csv')
del[df['Unnamed: 0.1']]
del[df['Unnamed: 0']]
df.columns

Index(['Name', 'state', 'summit_elev', 'vertical_drop', 'base_elev', 'trams',
       'fastEight', 'fastSixes', 'fastQuads', 'quad', 'triple', 'double',
       'surface', 'total_chairs', 'Runs', 'TerrainParks', 'LongestRun_mi',
       'SkiableTerrain_ac', 'Snow Making_ac', 'daysOpenLastYear', 'yearsOpen',
       'averageSnowfall', 'AdultWeekday', 'AdultWeekend', 'projectedDaysOpen',
       'NightSkiing_ac', 'cluster'],
      dtype='object')

## Create dummy features for categorical variables

**<font color='teal'> Create dummy variables for `State`. Add the dummies back to the dataframe and remove the original column for `State`. </font>**

In [457]:
dfo=df['state'] # select object type columns
dummy_state_values= pd.get_dummies(dfo)
dummy_state_values
df=pd.concat([df,dummy_state_values],axis=1)
del(df['state'])
df.head()


Name  summit_elev  ...  Wisconsin  Wyoming
0            Eaglecrest Ski Area         2600  ...          0        0
1               Hilltop Ski Area         2090  ...          0        0
2               Arizona Snowbowl        11500  ...          0        0
3            Sunrise Park Resort        11100  ...          0        0
4  Yosemite Ski & Snowboard Area         7800  ...          0        0

[5 rows x 61 columns]

## Standardize the magnitude of numeric features

**<font color='teal'> Using sklearn preprocessing, standardize the scale of the features of the dataframe except the name of the resort, which you don't need in the dataframe for modeling so it can be dropped here as well. You should  hold out your response variable(s) so that you can have their true values available for model performance review. Let's set `AdultWeekend` to the y variable as your response for scaling and modeling. Later, you will go back and consider the `AdultWeekday`, `dayOpenLastYear`, and `projectedDaysOpen`. For now leave, them in the development dataframe. </font>**

In [0]:
#from sklearn import preprocessing
#no_response_var_data=df.drop(['Name','AdultWeekend'], axis=1)
#y=df.AdultWeekend
#scaler = preprocessing.StandardScaler().fit(no_response_var_data)
#scaled=scaler.transform(no_response_var_data)

In [0]:
del(df['Name'])
y=df['AdultWeekend']
no_response_var_data=df.drop(df.loc[:,["AdultWeekend"]],axis=1)
scaled=preprocessing.StandardScaler().fit(no_response_var_data)
scaled
X_scaled=scaled.transform(no_response_var_data) 
#for col in no_response_var_data.columns:
#How do I scale?
#How do I go match the values to the orignial columns since I Just have a numpy array?



In [460]:
no_response_var_data

summit_elev  vertical_drop  base_elev  ...  West Virginia  Wisconsin  Wyoming
0           2600           1540       1200  ...              0          0        0
1           2090            294       1796  ...              0          0        0
2          11500           2300       9200  ...              0          0        0
3          11100           1800       9200  ...              0          0        0
4           7800            600       7200  ...              0          0        0
..           ...            ...        ...  ...            ...        ...      ...
290         9500           1000       8500  ...              0          0        1
291         7428            810       6619  ...              0          0        1
292         7808           1571       6237  ...              0          0        1
293         9663            990       8798  ...              0          0        1
294         9500           1100       8400  ...              0          0        1

[295 rows x 59 columns]

In [461]:
no_response_var_data.columns

Index(['summit_elev', 'vertical_drop', 'base_elev', 'trams', 'fastEight',
       'fastSixes', 'fastQuads', 'quad', 'triple', 'double', 'surface',
       'total_chairs', 'Runs', 'TerrainParks', 'LongestRun_mi',
       'SkiableTerrain_ac', 'Snow Making_ac', 'daysOpenLastYear', 'yearsOpen',
       'averageSnowfall', 'AdultWeekday', 'projectedDaysOpen',
       'NightSkiing_ac', 'cluster', 'Alaska', 'Arizona', 'California',
       'Colorado', 'Connecticut', 'Idaho', 'Illinois', 'Indiana', 'Iowa',
       'Maine', 'Maryland', 'Massachusetts', 'Michigan', 'Minnesota',
       'Missouri', 'Montana', 'Nevada', 'New Hampshire', 'New Jersey',
       'New Mexico', 'New York', 'North Carolina', 'Ohio', 'Oregon',
       'Pennsylvania', 'Rhode Island', 'South Dakota', 'Tennessee', 'Utah',
       'Vermont', 'Virginia', 'Washington', 'West Virginia', 'Wisconsin',
       'Wyoming'],
      dtype='object')

## Split into training and testing datasets

**<font color='teal'> Using sklearn model selectionm import train_test_split, and create a 75/25 split with the y = `AdultWeekend`. You will start by using the adult weekend ticket price as your response variable for modeling.</font>**

In [462]:
from sklearn.model_selection import train_test_split

#X_train, X_test, y_train, y_tes=train_test_split(no_response_var_data,y)
#scaled
y=y.ravel()
X_train, X_test, y_train, y_tes=train_test_split(X_scaled,y)
print(no_response_var_data.columns)


Index(['summit_elev', 'vertical_drop', 'base_elev', 'trams', 'fastEight',
       'fastSixes', 'fastQuads', 'quad', 'triple', 'double', 'surface',
       'total_chairs', 'Runs', 'TerrainParks', 'LongestRun_mi',
       'SkiableTerrain_ac', 'Snow Making_ac', 'daysOpenLastYear', 'yearsOpen',
       'averageSnowfall', 'AdultWeekday', 'projectedDaysOpen',
       'NightSkiing_ac', 'cluster', 'Alaska', 'Arizona', 'California',
       'Colorado', 'Connecticut', 'Idaho', 'Illinois', 'Indiana', 'Iowa',
       'Maine', 'Maryland', 'Massachusetts', 'Michigan', 'Minnesota',
       'Missouri', 'Montana', 'Nevada', 'New Hampshire', 'New Jersey',
       'New Mexico', 'New York', 'North Carolina', 'Ohio', 'Oregon',
       'Pennsylvania', 'Rhode Island', 'South Dakota', 'Tennessee', 'Utah',
       'Vermont', 'Virginia', 'Washington', 'West Virginia', 'Wisconsin',
       'Wyoming'],
      dtype='object')


Here, you'll start the actual modeling work. First let's fit a multiple linear regression model to predict the `AdultWeekend` price.

# Guided Capstone Step 5 - Modeling

You've cleaned and prepared the datasets, so now it's time to get into the most exciting part of this process: modeling! In this exercise, you will build three different models and compare each model's performance. In the end, you will choose the best model for proving insights to Big Mountain management.  

1.   Problem Identification 

2.   Data Wrangling 
  
3.   Exploratory Data Analysis 
 
4.   Pre-processing and Training Data Development

5.   **Modeling**
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

## Fit Models with Training Data Set

**<font color='teal'> Using sklearn, fit the model on your training dataset.</font>**

#### Model 1

In [0]:
#all first model set
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)

**<font color='teal'> Predict on the testing dataset and score the model performance with the y_test set and the y-pred values. The explained variance is a measure of the variation explained by the model. This is also known as the R-squared value. </font>**

In [0]:
from sklearn.metrics import mean_squared_error, r2_score

y_prediction=model.predict(X_test)
#print(y_prediction)
#print(model.score(X_test,y_tes))
#What is the difference between prediction and? Is it done correctly? And Score

In [465]:
explained_variance_score(y_prediction, y_tes)


-3.410649540569466e-11

In [466]:
mean_absolute_error(y_prediction, y_tes)

3039029325.537227

## Review Model Outcomes — Iterate over additional models as needed.

In [467]:
#adjust x columns, revmoe see if there is any impact
dropped = no_response_var_data.drop(['trams','fastEight','fastSixes','fastQuads','quad','triple','double'],axis=1)
X_train, X_test, y_train, y_tes=train_test_split(dropped,y)
model2 = lm.fit(X_train,y_train)
y_prediction=model2.predict(X_test)
print(model2.score(X_test,y_tes))

0.776976759397166


In [468]:
explained_variance_score(y_prediction, y_tes)

0.7696949566001378

In [469]:
mean_absolute_error(y_prediction, y_tes)

5.677011818059576

In [470]:
dropped2 = no_response_var_data.drop(['surface','total_chairs','Runs'],axis=1)
X_train, X_test, y_train, y_tes=train_test_split(dropped2,y)
model3 = lm.fit(X_train,y_train)
y_prediction=model3.predict(X_test)
print(model3.score(X_test,y_tes))

0.8021528112839117


**<font color='teal'> Print the intercept value from the linear model. </font>**

In [471]:
print(model.intercept_)


16.5183382065944


**<font color='teal'> The intercept is the mean `AdultWeekend` price for all the resorts given the other characteristics. The addition or subtraction of each of the coefficient values in the regression are numeric adjustments applied to the intercept to provide a particular observation's value for the `AdultWeekend` resulting value. Also, because you took the time to scale your x values in the training data, you can compare each of the coefficients for the features to determine the feature's importance. Print the coefficient values from the linear model and sort in descending order to identify the top ten most important features.</font>** 

Hint: make sure to review the absolute value of the coefficients, because the adjustment may be positive or negative, but what you are looking for is the magnitude of impact on your response variable.

In [472]:
print(X_train.columns)
coeff_df = pd.DataFrame(abs(model.coef_), X_train.columns, columns=['Coefficient'])  
coeff_df.sort_values(by=['Coefficient'],ascending=False).head(10)


Index(['summit_elev', 'vertical_drop', 'base_elev', 'trams', 'fastEight',
       'fastSixes', 'fastQuads', 'quad', 'triple', 'double', 'TerrainParks',
       'LongestRun_mi', 'SkiableTerrain_ac', 'Snow Making_ac',
       'daysOpenLastYear', 'yearsOpen', 'averageSnowfall', 'AdultWeekday',
       'projectedDaysOpen', 'NightSkiing_ac', 'cluster', 'Alaska', 'Arizona',
       'California', 'Colorado', 'Connecticut', 'Idaho', 'Illinois', 'Indiana',
       'Iowa', 'Maine', 'Maryland', 'Massachusetts', 'Michigan', 'Minnesota',
       'Missouri', 'Montana', 'Nevada', 'New Hampshire', 'New Jersey',
       'New Mexico', 'New York', 'North Carolina', 'Ohio', 'Oregon',
       'Pennsylvania', 'Rhode Island', 'South Dakota', 'Tennessee', 'Utah',
       'Vermont', 'Virginia', 'Washington', 'West Virginia', 'Wisconsin',
       'Wyoming'],
      dtype='object')


Coefficient
North Carolina    18.478746
West Virginia     13.367434
Maryland          10.860982
Montana            9.503155
Indiana            9.224856
Ohio               8.229622
Virginia           7.529982
Nevada             7.398096
South Dakota       6.652072
Idaho              6.282529

**<font color='teal'>You should see that the top ten important features are different states. However, the state is not something the managers at the Big Mountain Resort can do anything about. Given that you care more about actionable traits associated with ticket pricing, rebuild the model without the state features and compare the results. </font>**

#### Model 2

In [473]:
X_train.head()
X_train_no_states=X_train.drop(X_train.iloc[:,23:],axis=1)
X_test_no_states=X_test.drop(X_test.iloc[:,23:],axis=1)
print(X_train_no_states.columns)
model_no_states=lm.fit(X_train_no_states,y_train)
y_prediction=model_no_states.predict(X_test_no_states)
explained_variance_score(y_prediction, y_tes)



Index(['summit_elev', 'vertical_drop', 'base_elev', 'trams', 'fastEight',
       'fastSixes', 'fastQuads', 'quad', 'triple', 'double', 'TerrainParks',
       'LongestRun_mi', 'SkiableTerrain_ac', 'Snow Making_ac',
       'daysOpenLastYear', 'yearsOpen', 'averageSnowfall', 'AdultWeekday',
       'projectedDaysOpen', 'NightSkiing_ac', 'cluster', 'Alaska', 'Arizona'],
      dtype='object')


0.7750688886074415

In [474]:
mean_absolute_error(y_prediction, y_tes)

6.39676281843864

In [475]:
coeff_df = pd.DataFrame(abs(model_no_states.coef_), X_train_no_states.columns, columns=['Coefficient'])  
coeff_df.sort_values(by=['Coefficient'],ascending=False).head(10)

Coefficient
Alaska            3.638330
cluster           2.206199
trams             2.074306
quad              1.331244
Arizona           1.144750
fastSixes         0.929711
AdultWeekday      0.796800
fastQuads         0.657727
LongestRun_mi     0.413542
triple            0.371608

**<font color='teal'> When reviewing your new model coefficients, note that `summit_elev` is now in the number two spot. From a managerial perspective, this is also difficult to change and highly correlated with `base_elev` and `vertical_drop`.  This time rebuild the model without the state features and without the `summit_elev` and without `base_elev`and compare the results. </font>**

#### Model 3

In [476]:
print(X_test_no_states.columns)
X_train_no_states_no_elev=X_test_no_states.drop(["summit_elev","base_elev"],axis=1)

X_test_no_states_no_elev=X_test_no_states.drop(["summit_elev","base_elev"],axis=1)
X_test_no_states_no_elev.columns

model_no_states_no_elev=lm.fit(X_test_no_states_no_elev,y_tes)
y_prediction=model_no_states.predict(X_test_no_states_no_elev)
explained_variance_score(y_prediction, y_tes)

Index(['summit_elev', 'vertical_drop', 'base_elev', 'trams', 'fastEight',
       'fastSixes', 'fastQuads', 'quad', 'triple', 'double', 'TerrainParks',
       'LongestRun_mi', 'SkiableTerrain_ac', 'Snow Making_ac',
       'daysOpenLastYear', 'yearsOpen', 'averageSnowfall', 'AdultWeekday',
       'projectedDaysOpen', 'NightSkiing_ac', 'cluster', 'Alaska', 'Arizona'],
      dtype='object')


0.8315965057377195

In [477]:
coeff_df = pd.DataFrame(abs(model_no_states_no_elev.coef_), X_test_no_states_no_elev.columns, columns=['Coefficient'])  
coeff_df.sort_values(by=['Coefficient'],ascending=False).head(10)

Coefficient
trams                5.121645
fastSixes            4.854968
quad                 2.549204
LongestRun_mi        2.487726
fastQuads            1.958132
triple               1.889541
cluster              1.703060
double               1.614285
AdultWeekday         0.658791
daysOpenLastYear     0.135043

In [478]:
mean_absolute_error(y_prediction, y_tes)

5.405287525836947

## Identify the Final Model

**<font color='teal'> Review the model performances in the table below and choose the best model for proving insights to Big Mountain management about what features are driving ski resort lift ticket prices. Type your choice in the final markdown cell. You will explain your selection during the next step of the guided casptone. </font>**

| Model | Explained Variance| Mean Absolute Error|Features Dropped|
| --- | --- | --- | --- |
| Model 1. | -3.410649540569466 |3039029325.537227  |-|
| Model 2. | 0.7750688886074415|6.39676281843864 |'state'|
| Model 3. | 0.8315965057377195 |5.405287525836947 |'state','summit_elev','base_elev'|

Model Selection:

In [0]:
#MOdel 3